pymc/distributions/multivariate.py

```diff
@@ -289,10 +293,72 @@ def logp(value, mu, cov):
         return check_parameters(
             norm - 0.5 * quaddist - logdet,
             ok,
-            msg="posdef",
+            msg="posdef covariance",
         )
 
 
+class PrecisionMvNormalRV(SymbolicRandomVariable):
+    r"""A specialized multivariate normal random variable defined in terms of precision.
+
+    This class is introduced during specialization logprob rewrites, and not meant to be used directly.
+    """
+
+    name = "precision_multivariate_normal"
+    extended_signature = "[rng],[size],(n),(n,n)->(n)"
+    _print_name = ("PrecisionMultivariateNormal", "\\operatorname{PrecisionMultivariateNormal}")
+
+    @classmethod
+    def rv_op(cls, mean, tau, *, rng=None, size=None):
+        rng = normalize_rng_param(rng)
+        size = normalize_size_param(size)
+        cov = pt.linalg.inv(tau)
+        next_rng, draws = multivariate_normal(mean, cov, size=size, rng=rng).owner.outputs
+        return cls(
+            inputs=[rng, size, mean, tau],
+            outputs=[next_rng, draws],
+        )(rng, size, mean, tau)
+
+
+@_logprob.register
+def precision_mv_normal_logp(op: PrecisionMvNormalRV, value, rng, size, mean, tau, **kwargs):
+    [value] = value
+    k = value.shape[-1]
+    delta = value - mean
+    det_sign, logdet_tau = pt.nlinalg.slogdet(tau)
```

Comment:
I'm not sure if the slogdet is necessarily a good idea here. Internally at least in numpy this does a lu factorization, which is I think takes theoretically about twice as long as the cholesky, and should usually be less stable. (But I think the performance can differ a lot based on number of threads and blas). So for a matrix that is not constant this might be slower than the usual MvNormal right now.

So I think it is better to use the cholesky decomposition here and use that to get the logdet.

Reply:
In the example above the matrix is not constant, so you can use it to benchmark

Reply:
Here: https://github.com/pymc-devs/pymc/pull/7345#issuecomment-2182930220

Reply:
And it twas 2x faster logp and 100x faster dlogp on my crappy PC. Are you skeptical of those numbers?

Reply:
Are you suggesting something like this https://math.stackexchange.com/questions/2001041/logarithm-of-the-determinant-of-a-positive-definite-matrix ?

Reply:
Yes, `logdet = -2*pt.diag(cholesky(inv_cov)).sum()`.

The benchmark code uses a constant tau though, or am I missing something?

Reply:
No it has a shared variable which is not constant for PyTensor.

I set the logp to 
```python
    [value] = value
    k = value.shape[-1]
    delta = value - mean
    logdet_tau = 2 * pt.trace((pt.log(pt.linalg.cholesky(tau))), axis1=-2, axis2=-1)
    # det_sign, logdet_tau = pt.nlinalg.slogdet(tau)
    quadratic_form = delta.T @ tau @ delta
    logp = -0.5 * (k * pt.log(2 * pt.pi) - logdet_tau + quadratic_form)
    return check_parameters(
        logp,
        # det_sign > 0,
        logdet_tau > 0,
        msg="posdef precision",
    )
```

And then ran this:
```python
%env OMP_NUM_THREADS=1

import pymc as pm
import numpy as np
import pytensor
import pytensor.tensor as pt

rng = np.random.default_rng(123)

n = 100
L = rng.uniform(low=0.1, high=1.0, size=(n,n))
Sigma = L @ L.T
Q_test = np.linalg.inv(Sigma)
x_test = rng.normal(size=n)

with pm.Model(check_bounds=False) as m:
    Q = pm.Data("Q", Q_test)
    x = pm.MvNormal("x", mu=pt.zeros(n), tau=Q)

logp_fn = m.compile_logp().f
logp_fn.trust_input=True
print("logp")
pytensor.dprint(logp_fn)

dlogp_fn = m.compile_dlogp().f
dlogp_fn.trust_input=True
print("dlogp")
# pytensor.dprint(dlogp_fn)

np.testing.assert_allclose(logp_fn(x_test), np.array(-1789.93662205))

np.testing.assert_allclose(np.sum(dlogp_fn(x_test) ** 2), np.array(18445204.8755109), rtol=1e-6)

# With slogdet: 318 µs ± 42.5 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# With cholesky logdet: 310 µs ± 35.7 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit -n 10000 logp_fn(x_test)

# Before: 2.45 ms
# With slogdet: # 31 µs ± 2.43 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# With cholesky logdet: 32.5 µs ± 4.51 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit -n 10000 dlogp_fn(x_test)
```
At least for this example it's pretty mach a draw on my machine. Do you get different results?

Reply:
the logdet > 0 thing is silly haha, but anyway we are running without check_bounds so it doesn't show in the graph

Reply:
Here would be some code with a non-constant matrix, but I get a NotImplementederror for the grad of Blockwise?

```
import pymc as pm
import numpy as np
import pytensor
import pytensor.tensor as pt

rng = np.random.default_rng(123)

n = 100
L = rng.uniform(low=0.1, high=1.0, size=(n,n))
Sigma = L @ L.T
Q_test = np.linalg.inv(Sigma)
x_test = rng.normal(size=n)
v_test = rng.normal(size=n)

with pm.Model(check_bounds=False) as m:
    Q = pm.Data("Q", Q_test)
    v = pm.Normal("v", shape=n)
    x = pm.MvNormal("x", mu=pt.zeros(n), tau=Q + v[None, :] * v[:, None])

logp_fn = m.compile_logp().f
logp_fn.trust_input=True
print("logp")
#pytensor.dprint(logp_fn)


dlogp_fn = m.compile_dlogp().f
dlogp_fn.trust_input=True
print("dlogp")
#pytensor.dprint(dlogp_fn)

#np.testing.assert_allclose(logp_fn(x_test, v_test), np.array(-1789.93662205))

#np.testing.assert_allclose(np.sum(dlogp_fn(x_test, v_test) ** 2), np.array(18445204.8755109), rtol=1e-6)
with threadpoolctl.threadpool_limits(1):

    # Before: 2.66 ms
    # After: 1.31 ms
    %timeit logp_fn(x_test, v_test)
    
    # Before: 2.45 ms
    # After: 72 µs
    %timeit dlogp_fn(x_test, v_test)
```

Reply:
Your expression moves the log above the diagonal, but I think the minus in `-2 * ` is wrong?

Reply:
The performance of the original code looks I think pretty bad by the way:
```
#353 μs ± 4.53 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
#1.61 ms ± 32.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
If everything is implemented well I don't think there is a good reason why the gradient should be much slower than just the logp. Factoring the matrx onces should be enough.
I guess this is because we compute the pullback of the cholesky or so, even though the pullback of the logdet is actually pretty easy if you already have a factorization...
There might be a nice usecase for an OpFromGraph and overwriting the forward values hiding here somewhere...

Reply:
Seems to be 40us (1.2x faster) with the cholesky now that I only do log on the diagonal. Funny enough it's fusing the log and the sum, which is our only elemwise + reduce fusion we have :)

```python
logdet_tau = 2 * pt.log(pt.diagonal(pt.linalg.cholesky(tau), axis1=-2, axis2=-1)).sum()
```

```python
%env OMP_NUM_THREADS=1

import pymc as pm
import numpy as np
import pytensor
import pytensor.tensor as pt

rng = np.random.default_rng(123)

n = 100
L = rng.uniform(low=0.1, high=1.0, size=(n,n))
Sigma = L @ L.T
Q_test = np.linalg.inv(Sigma)
x_test = rng.normal(size=n)

with pm.Model(check_bounds=False) as m:
    Q = pm.Data("Q", Q_test)
    x = pm.MvNormal("x", mu=pt.zeros(n), tau=Q)

logp_fn = m.compile_logp().f
logp_fn.trust_input=True
print("logp")
pytensor.dprint(logp_fn)

dlogp_fn = m.compile_dlogp().f
dlogp_fn.trust_input=True
print("dlogp")
pytensor.dprint(dlogp_fn)

np.testing.assert_allclose(logp_fn(x_test), np.array(-1789.93662205))

np.testing.assert_allclose(np.sum(dlogp_fn(x_test) ** 2), np.array(18445204.8755109), rtol=1e-6)

# With slogdet: 236 µs ± 31.7 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# With cholesky logdet: 192 µs ± 25.6 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit -n 10000 logp_fn(x_test)

# With slogdet: 29.8 µs ± 3.19 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# With cholesky logdet: 32.5 µs ± 4.51 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit -n 10000 dlogp_fn(x_test)
```

Reply:
Pushed a commit with the cholesky factorization